binary-search-tree/node-js/inventory-management.js

```diff
@@ -18,16 +18,37 @@ class BinarySearchTree {
         this.root = null;
     }
 
-    search(id) {
-        console.error("you must implement search");
+    search(id, currentNode) {
+        if (id === null) {
+            return null;
+        } else {
+            return this.searchRecursive(id, currentNode)
+        }
     }
-
+        // console.error("you must implement search");
+        // 1. Start from the root node
+        // 2. Compare the current node with the search value. Return early if found
+        // 3. Determine if we should go left or right
+        // 4. Update the current node based on the decision from (3). Rinse and repeat from (2)
+    searchRecursive(id, currentNode) {
```

Comment:
I urge you to consider the thread of execution of this method when `currentNode` is a leaf node (has no children) and doesn't have the same `id` as the search value.

Here's such a thread of execution

1. `searchRecursive` is called with `id=9` and `currentNode` is a leaf node with `val.id=5`.
2. The first `if` statement is not true since `currentNode.val.id` is not equal to `id`. So we move to the next `else if` statement.
3. The second `else if` statement is true since `id` is greater than `currentNode.val.id` (9 > 5). So we call `searchRecursive` again with `id=9` and `currentNode=currentNode.right`.
   - This creates a new stack frame on the call stack with a new set of parameters: `id=9` and `currentNode` is `null` because `currentNode.right` is `null`.
   - The function then enters the new stack frame with these new parameters and the thread of execution continues from the beginning of the function with the new set of parameters.
   - Since `currentNode` is `null`, the next conditional statement `if(currentNode.val.id === id )` evaluates to `null.val.id === id` which is trying to access the `val` property of `null`, resulting in the error "TypeError: Cannot read property `val` of null."


Being able to follow the thread of execution is important!

Reply:
sounds good, thanks for the feedback, am on it already